Apps/TimelineApp/core/sensor/src/main/java/uk/ac/cam/cares/jps/sensor/source/worker/SensorUploadWorker.java

```diff
@@ -0,0 +1,127 @@
+                              SensorNetworkSource sensorNetworkSource,
+                              SensorLocalSource sensorLocalSource) {
+        super(context, workerParams);
+        this.sensorNetworkSource = sensorNetworkSource;
+        this.sensorLocalSource = sensorLocalSource;
+        deviceId = workerParams.getInputData().getString("deviceId");
+        String selectedSensorsJson = workerParams.getInputData().getString("selectedSensors");
+        selectedSensors = new ArrayList<>();
+
+        try {
+            JSONArray jsonArray = new JSONArray(selectedSensorsJson);
+            for (int i = 0; i < jsonArray.length(); i++) {
+                String sensorName = jsonArray.getString(i);
+                selectedSensors.add(SensorType.valueOf(sensorName));
+            }
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @NonNull
+    @Override
+    public Result doWork() {
+        // Retrieve data and upload
+        try {
+            uploadSensorData();
+
+            return Result.success();
+        } catch (Exception e) {
+            Log.e("DataUploadWorker", "Error uploading sensor data", e);
+            return Result.failure(); // data should be sent to local alr
+        }
+    }
+
+    private void uploadSensorData() throws JSONException, IOException {
+        int PAGE_SIZE = 1500;
+        int offset = 0;
+        boolean hasMoreData = true;
+
+        while (hasMoreData) {
+            JSONArray allSensorData = sensorLocalSource.retrieveUnUploadedSensorData(selectedSensors, PAGE_SIZE, offset);
+            Log.i("DataUploadWorker", "Retrieved " + allSensorData.length() + " items from local storage.");
+
+            if (allSensorData.length() < PAGE_SIZE) {
+                hasMoreData = false;
+            }
+
+            String jsonString = allSensorData.toString();
+            byte[] compressedData = compressData(jsonString);
+
+            // Send to the network
+            if (allSensorData.length() > 0) {
+                LOGGER.info("Attempting to send " + allSensorData.length() + " items to the network.");
+                LOGGER.info("All sensor data " + allSensorData);
+                sensorNetworkSource.sendPostRequest(deviceId, compressedData, allSensorData);
+                LOGGER.info("Accumulated data sent to network.");
+            } else {
+                LOGGER.info("No accumulated data to send to the network.");
+            }
+            offset += PAGE_SIZE;
+        }
+
+
+    }
+
+    /**
+     * Compresses a given string into a GZIP-compressed byte array.
+     * This method uses the GZIP compression algorithm to reduce the size of the input string.
+     *
+     * @param data The input string to be compressed.
+     * @return A byte array containing the GZIP-compressed data.
+     * @throws IOException If an I/O error occurs during the compression process.
+     */
+    public static byte[] compressData(String data) throws IOException {
```

Comment:
this function has also been defined in SensorService. Move it to some util class as static method